src/sinks/util/buffer/compression.rs

```diff
@@ -0,0 +1,144 @@
+use serde::{de, ser};
+use std::fmt;
+
+#[derive(Debug, Derivative, Copy, Clone, Eq, PartialEq)]
+#[derivative(Default)]
+pub enum Compression {
+    #[derivative(Default)]
+    None,
+    Gzip(usize),
+}
+
+impl Compression {
+    pub fn default_gzip() -> Compression {
+        Compression::Gzip(6)
+    }
+
+    pub fn content_encoding(&self) -> Option<&'static str> {
+        match self {
+            Self::None => None,
+            Self::Gzip(_) => Some("gzip"),
+        }
+    }
+
+    pub fn extension(&self) -> &'static str {
+        match self {
+            Self::None => "log",
+            Self::Gzip(_) => "log.gz",
+        }
+    }
+}
+
+#[cfg(feature = "rusoto_core")]
+impl From<Compression> for rusoto_core::encoding::ContentEncoding {
+    fn from(compression: Compression) -> Self {
+        match compression {
+            Compression::None => rusoto_core::encoding::ContentEncoding::Identity,
+            Compression::Gzip(level) => {
+                rusoto_core::encoding::ContentEncoding::Gzip(None, level as u32)
+            }
+        }
+    }
+}
+
+impl<'de> de::Deserialize<'de> for Compression {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
```

Comment:
I wonder if this couldn't be handled automatically by serde with a flattened enum, something like the following (untested):
```
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
enum CompressionName {
    None,
    Gzip,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(untagged)]
enum CompressionConfig {
    Name(String),
    NameLevel { codec: String, level: String },
}

impl TryInto<Compression> for CompressionConfig { … }
```

Reply:
I think it can, but I still will need to verify the decoded config in `TryInto`... so I think it's fine to go with custom Deserializer